File: 0128-longest-consecutive-sequence/0128-longest-consecutive-sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # approach: hashset of starts of sequences, then iterate from each start
        
        if not nums:
            return 0
        
        numset = set(nums)
        starts = []
        longest = 0
        
        for num in nums:
            if num - 1 not in numset:
                starts.append(num)
                
        for start in starts:
            num = start
            length = 1
            
            while (num + 1 in numset):
                length += 1
                num = num + 1
        
            longest = max(longest, length)
        
        return longest

File: 0128-longest-consecutive-sequence/0128-longest-consecutive-sequence.java
class Solution {
    public int longestConsecutive(int[] nums) {
        // approach: two pass array, one to find starts of sequences and one to
        // find longest sequence
        
        int result = 0;
        
        Set<Integer> hashset = new HashSet<>();
        List<Integer> starts = new ArrayList<>();
        
        for (int num : nums) {
            hashset.add(num);
        }
        
        for (int num : nums) {
            if (hashset.contains(num - 1)) {
                continue;
            }
            
            starts.add(num);
        }
        
        for (int start : starts) {
            int length = 1;
            int num = start;
            
            while (hashset.contains(num + 1)) {
                length++;
                num++;
            }
            
            result = Math.max(result, length);
        }
        
        return result;
    }
}